### DISCUSSION TITLE
Binding multiple values into a single type.

### DISCUSSION THREAD

**Discussion by robcao** [spectreconsole/spectre.console#1611] [Q&A]
I'm looking for advice on the best way to bind multiple input from the command line into a single object.

I have a fairly complex object. Instead of passing in values for the object directly through the command line, I want to be able to pass in the location to one or more yaml files containing the object.

I have implemented this as follows.

```csharp
/// <summary>
/// Command settings for the management of a user.
/// </summary>
internal class UserManagementCommandSettings : BaseCommandSettings
{
	[CommandOption("--meta <VALUES>")]
	[TypeConverter(typeof(YamlFileConverter<UserMeta>))]
	[Description("Yaml file(s) containing the user metadata. If multiple files are specified, later files will take precedence.")]
	public UserMeta Meta { get; set; } = new();
}

/// <summary>
/// Generic type converter that takes in either a single yaml file or multiple yaml files, an in-memory model to the contents of the file(s).
/// </summary>
/// <remarks>
/// If multiple files are specified, the precedence is in reverse order.
/// </remarks>
internal class YamlFileConverter<T> : TypeConverter where T : new()
{
	public override object ConvertFrom(ITypeDescriptorContext? context, CultureInfo? culture, object value)
	{
                IConfigurationBuilder Builder { get; init; } = new ConfigurationBuilder();
		if (value is string path)
		{
			Builder.AddYamlFile(path, optional: false);
		}

		else if (value is IEnumerable<string> paths)
		{
			foreach (string next in paths)
			{
				Builder.AddYamlFile(next, optional: false);
			}
		}

		else
		{
			throw new NotSupportedException("Must be either a file path or a collection of file paths.");
		}

		IConfiguration root = Builder.Build();

		T model = new();

		root.Bind(model);

		return model;
	}
}

```

This works exactly as desired when I pass in a single file, eg `my-cli --meta file-1.yaml`.

However, when I pass in multiple files, eg `my-cli --meta file-1.yaml --meta file-2.yaml`, the `ConvertFrom` method is invoked twice, and only the value of the second iteration is being kept.

I can get around this by making the converter stateful like below, but I am noticing that a single `TypeConverter<T>` instance is being re-used for all conversions.

```csharp

/// <summary>
/// Generic type converter that takes in either a single yaml file or multiple yaml files, an in-memory model to the contents of the file(s).
/// </summary>
/// <remarks>
/// If multiple files are specified, the precedence is in reverse order.
/// </remarks>
internal abstract class YamlFileConverter<T> : TypeConverter where T : new()
{
	/// <summary>
	/// The <see cref="IConfiguration"/>
	/// </summary>
	/// <remarks>
	/// </remarks>
	public IConfigurationBuilder Builder { get; init; } = new ConfigurationBuilder();

	public override object ConvertFrom(ITypeDescriptorContext? context, CultureInfo? culture, object value)
	{
		if (value is string path)
		{
			Builder.AddYamlFile(path, optional: false);
		}

		else if (value is IEnumerable<string> paths)
		{
			foreach (string next in paths)
			{
				Builder.AddYamlFile(next, optional: false);
			}
		}

		else
		{
			throw new NotSupportedException("Must be either a file path or a collection of file paths.");
		}

		IConfiguration root = Builder.Build();

		T model = new();

		root.Bind(model);

		return model;
	}
}
```

For example, if I alter the settings to have more than one property using the same type converter, and then run `my-cli --meta file-1.yaml --meta file-2.yaml --meta2 file-3.yaml`, because the above converter is stateful, all 3 files are being merged together.

Is there a better way to achieve this? Instead of using `<VALUES>`, should I accept a single delimited list of file paths, and do the parsing myself? Ideally, I would like to avoid any sort of correctness requirements based on the statefulness of the type converter.

```csharp
/// <summary>
/// Command settings for the management of a user.
/// </summary>
internal class UserManagementCommandSettings : BaseCommandSettings
{
	[CommandOption("--meta <VALUES>")]
	[TypeConverter(typeof(YamlFileConverter<UserMeta>))]
	[Description("Yaml file(s) containing the user metadata. If multiple files are specified, later files will take precedence.")]
	public UserMeta Meta { get; set; } = new();

        [CommandOption("--meta2")]
	[TypeConverter(typeof(YamlFileConverter<UserMeta>))]
	public UserMeta Meta2 { get; set; } = new();
}
```

**Comment by elgonzo** [comment]
> Is there a better way to achieve this? 

Given the Spectre.Console.Cli functionality, i would stick with a [string array option](https://spectreconsole.net/cli/settings#option-arrays) and then do the conversion into a `UserMeta` instance (as a separate property) in the [Validate method](https://spectreconsole.net/cli/settings#validation).

Your struggles with the type converter not withstanding, I would _not_ recommend using a type converter via the `[TypeConverter]` attribute for this particular use case, because you won't be easily able to generate meaningful user-facing errors such as a YAML file not existing or having the wrong format/layout/structure/etc.
If your type converter would throw (for example a FileNotFoundException), all the user sees is an error message that the given file path couldn't be converted to a UserMeta instance without giving a concrete clue about the reason of the error.

On the other hand, exceptions thrown in the Validate method will result in a user-facing error with the exception message. Even better, the Validate method allows you to produce a user-facing error message without needing to rely on exceptions to carry user-facing errors ([as shown in the example here](https://spectreconsole.net/cli/settings#validation)).

**Comment by robcao** [comment]
Thanks! I'll play around with it. Your response is greatly appreciated.
